### ISSUE TITLE
[bug] android release build fails with linker error

### ISSUE THREAD

**Issue by kicsiede** [ChasakisD/Xamarin.Forms.BackgroundKit#47]
hello,

when attempting a release build on Android i get the following linker error, even though i'm not using MaterialCard, and do invoke XamarinBackgroundKit.Android.BackgroundKit.Init();

iOS builds fine.

```
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018: The "LinkAssemblies" task failed unexpectedly.
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018: Mono.Linker.MarkException: Error processing method: 'System.Void XamarinBackgroundKit.Android.Renderers.MaterialCardRenderer::.ctor(Android.Content.Context)' in assembly: 'XamarinBackgroundKit.Android.dll' ---> Mono.Cecil.ResolutionException: Failed to resolve Xamarin.Forms.Material.Android.MaterialContextThemeWrapper Xamarin.Forms.Material.Android.MaterialContextThemeWrapper::Create(Android.Content.Context)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.HandleUnresolvedMethod(MethodReference reference)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.MarkMethod(MethodReference reference)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.MarkInstruction(Instruction instruction)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.MarkMethodBody(MethodBody body)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.ProcessQueue()
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    --- End of inner exception stack trace ---
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.ProcessQueue()
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.Process()
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at MonoDroid.Tuner.MonoDroidMarkStep.Process(LinkContext context)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Pipeline.Process(LinkContext context)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at MonoDroid.Tuner.Linker.Process(LinkerOptions options, ILogger logger, LinkContext& context)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Xamarin.Android.Tasks.LinkAssemblies.Execute(DirectoryAssemblyResolver res)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Xamarin.Android.Tasks.LinkAssemblies.Execute()
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()

```

**Comment by ChasakisD** [comment]
Add FormsMaterial.Init() and test again

**Comment by kicsiede** [comment]
it is there as well

**Comment by ChasakisD** [comment]
The problem is with Visual Material. What version of Visual Material are you using?

**Comment by kicsiede** [comment]
v3.6.0.539721 (both XF & XF Visual)

interestingly iOS has not problem with it

**Comment by ChasakisD** [comment]
The problem is with MaterialContextThemeWrapper that exists on Visual.Material. I faced the same issue before on production applications without this library. Try update to Xamarin.Forms 4.x and see if the problem persists.

**Comment by kicsiede** [comment]
i can give it a try, but won't be able to go in production with 4 :(

* 4.0 builds indeed.

**Comment by ChasakisD** [comment]
Why you cannot use 4.x builds in production? Try 4.1, and then i will try to replicate the issue.

**Comment by kicsiede** [comment]
honest answer: no time/capacity to fight another shitload of newly introduced bugs in 4.x :D

4.1 doesn't build either.  but different error ofc. (I don't think this is related to your component)

```
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018: The "LinkAssemblies" task failed unexpectedly.
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018: Mono.Linker.MarkException: Error processing method: 'System.Void AndroidBug.App::InitializeComponent()' in assembly: 'AndroidBug.dll' ---> Mono.Cecil.ResolutionException: Failed to resolve System.Void Xamarin.Forms.Internals.ResourceLoader/ResourceLoadingQuery::set_Instance(System.Object)
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2129,5): error MSB4018:    at Mono.Linker.Steps.MarkStep.HandleUnresolvedMethod(MethodReference reference)

```


why do you have Preserve.All though? i'm not even using materialcard, so the the linker could strip it.

**Comment by ChasakisD** [comment]
Try Clean the solution, (Maybe delete bin, obj will help) and then try build in release. I have used this library in numerous projects with 4.1 X.F. and above and build in release works fine.

**Comment by kicsiede** [comment]
the problem is Preserve.AllMembers. why do you have that? it just forces the linker to try and link all members even if they are unused.
of course i can turn off linking so that it builds, but then i get a 100mb apk. :)

if i do want to preserve some members, i can use a link description file explicitly.

**Comment by ChasakisD** [comment]
This will solve part of the problem. The only thing that can be cut here is the MaterialCardRenderer. And if the linker will not link Visual.Material a lot of functionality that this library provides will not work because they depend on that. If the problem is on MaterialContextWrapper, use Visual="Material" on a Frame and you will get the exact same error. The solution is not to just let it build. 

Try add the following line to MainActivity or exclude Xamarin.Forms.Material.Android.MaterialContextThemeWrapper from linking and see if it builds
```cs
var unused = typeof(MaterialContextThemeWrapper);
```

**Comment by kicsiede** [comment]
Sorry I can't follow your thought process, and I don't really get what your point is here. Obviously I do have (direct) Visual.Material dependency in my projects, so it is the forced MaterialContextWrapper dependency that causes the problem.

So, to reiterate:
- I have a perfectly valid solution, that uses XF v3.6, XF Visual v3.6, and MaterialContentView v1.2.3 from your library.
- It builds perfectly with and without linking on iOS
- It builds without linking on Android (but the apk size is +60mb more this way, because other dependencies aren't linked either)

I don't use MaterialCard which I guess has a dependency from the XF Visual 4+ assembly on Android. (* meanwhile i tried it, and indeed it doesn't work at all on Android)
So being able to cut MaterialCardRenderer is exactly what would solve the problem here.

Are you saying that the solution is to either go with a +60mb apk size, or build my own version without Preserve, assuming changing the XF version number is not possible?

**Comment by ChasakisD** [comment]
Just to make it clear,

The problem is that the linker is cutting out MaterialContextThemeWrapper, not that is forced to the project. This class is used to every renderer in Visual.Material on Android. The reason I told you to update on 4.1 was that this issue with the linker cutting out that class was resolved in 4.1 release. I did not assume that you are doing something wrong and of course to DISABLE linking.

The problem here is not with the library, but with Visual.Material, because the MaterialContextThemeWrapper is not preserved, thus MaterialCardRenderer fails. Material Card has zero dependencies to Xamarin 4+. It just needs a class that the linker is not preserving.

There are 2 possible solutions over here.

The first one is to manually preserve MaterialContextThemeWrapper. This can be done either by holding a fake dependency to it like:
```cs
var unused = typeof(MaterialContextThemeWrapper);
```
or by exclude that class from linking in a linker config. No need to disable the linking, just to make sure that MaterialContextThemeWrapper will compile. This will not end with a 60mb apk size.

The second is one is to let linker link MaterialCardRenderer. By doing that, every user that uses MaterialCard and has the linking enabled, will probably get a runtime exception, because the MaterialContextThemeWrapper will not be found.

To make it even more clear,

> Obviously I do have (direct) Visual.Material dependency in my projects, so it is the forced MaterialContextWrapper dependency that causes the problem.  

It is not forced in, it just not preserved from linker   

>  So being able to cut MaterialCardRenderer is exactly what would solve the problem here.

It will hide, not solve, the problem

To summarize, try the first solution with X.F 3.6 and let me know.

**Comment by kicsiede** [comment]
first of all, the linker removes unused references. so it is impossible that you have a reference to the renderer, but you dont have a reference to the ThemeWrapper, because the renderer has a reference to the themewrapper. (to be more precise you must have a reference to the constructor of the renderer to make any use of it, and since it instatly makes a reference to the themewrapper, it wont be removed:) 
thats exactly what the linker does, goes through these dependencies, and keeps only those that it visited. works kind of like the garbage collector.
the only problem appears for example when you have eventhandlers that are referenced from xaml. these references the linker cannot have knowledge of. and for these kind of cases are the preserve attribute, to mark those methods/members that have no apparent callers, but are accessed through other means. so runtime errors only happen when you blindly call some unreferenced method through reflection assuming it wasnt stripped. (or you silently realise that you dont have the actual method and not even try to call it.)

on the other hand if you dont link, you will get a runtime error, and only when try to load the actual assembly that is missing a referenced method. thats why if you try a debug build with your lib with XF v3.6 and place a MaterialCard on the form you'll get a runtime error on Android when starting up.

strangely enough I've managed to get it working with 3.6 by rebuilding your lib to use XF 3.6. i figured 3.5 indeed doesnt have the wrapper class yet, but 3.6 does, so tried upgrading the nuspec to 3.6, but it didnt make a change. then rebuilt your whole lib to reference v3.6 all the way, and all of a sudden the linker error disappeared. i could even use the MaterialCard! which was weird. but then i get the linker error when trying to link against XF 4.1... so it must be some binary incompatibility issue between v3.6 and 4, or i dont know...

**Comment by bentmar** [comment]
having the exact same issue on xf 3.6. 
on 3.6 because of the same reason as @kicsiede (I couldnt have said it better myself). 

@ChasakisD  could you provide a nuget thats based on 3.6?  :)

**Comment by ChasakisD** [comment]
@bentmar I will upload soon a new version without the need of ContextWrapper

**Comment by ChasakisD** [comment]
Fixed by #50
